### PR TITLE
Serialization to produce registerable entities

### PR DESCRIPTION
**Pull request by wild-endeavor**
# TL;DR
This is step 1 of the issue linked below.  This corrects the serialization component of pyflyte.

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [x] Unit tests added
 - [x] Code documentation added
 - [x] Any pending items have an associated Issue

## Complete description
* Minor update to the registration code to just aggregate everything into a list first, and then run register.
* This PR basically copies the registration process, but instead of registering against admin, will create the objects necessary for registration at a later time.
* The same topo sort is used to correctly ascertain the order in which things should be registered later, enforced by adding an index to the start of the filename.
* This code path does not rely on the engine.

## Tracking Issue
https://github.com/lyft/flyte/issues/225

## Follow-up issue
NA

### CODE REVIEW

File: flytekit/common/workflow.py [hunk 1]
```diff
@@ -286,6 +287,20 @@ def register(self, project, domain, name, version):
             self._id = old_id
             raise
 
+    @_exception_scopes.system_entry_point
+    def serialize(self):
```
Comment (matthewphsmith) [inline flytekit/common/workflow.py:291]: if this is going to be a required method on each registerable object, let's add it as an abstract method to the RegisterableEntity mixin.

File: flytekit/common/workflow.py [hunk 2]
```diff
@@ -286,6 +287,20 @@ def register(self, project, domain, name, version):
...
+        """
+        Serializing a workflow should produce an object similar to what the registration step produces, in preparation
+        for actual registration to Admin.
+
+        :rtype: flyteidl.admin.workflow_pb2.WorkflowSpec
```
Comment (matthewphsmith) [inline flytekit/common/workflow.py:296]: Also, if we're going to serialize, maybe go straight to bytes and call `SerializeToString()`?

Comment (matthewphsmith) [reply flytekit/common/workflow.py:296]: And if we want to keep it as the IDL object, let's refactor the engine to call this method so we don't have duplicate code.

File: flytekit/clis/sdk_in_container/serialize.py [hunk 3]
```diff
@@ -2,49 +2,120 @@
...
+            project,
+            domain,
+            name,
+            version
+        )
+        loaded_entities.append(o)
+
+    i = 0
+    zero_padded_length = _determine_text_chars(len(loaded_entities))
+    for entity in loaded_entities:
```
Comment (matthewphsmith) [inline flytekit/clis/sdk_in_container/serialize.py:92]: nit: `for i, entity in enumerate(..):`

File: flytekit/common/workflow.py (hunk 2 above)
Comment (wild-endeavor) [reply flytekit/common/workflow.py:296]: I'm gonna go with option 2 here.

File: flytekit/common/workflow.py (hunk 1 above)
Comment (wild-endeavor) [reply flytekit/common/workflow.py:291]: good point

File: flytekit/common/workflow.py (hunk 2 above)
Comment (wild-endeavor) [reply flytekit/common/workflow.py:296]: Actually, can I just leave it as is?  It's not quite the same (serialize calls to_flyteidl already).

So my plan is to remove the register call completely in a future PR (since registration happens from a container, we can control this, and only will do this after all refactorator PRs are done, or possibly rewrite how pyflyte register itself works).

But I also don't want bytes quite yet.  I haven't quite worked out how it'll work but I want the option to serialize these all to a pipe.  To do this, the serialize command will also need to emit the size of each object (since protobuf doesn't have a native delimiter).  I might try to do that then.  If i change it now then i can't use the current `write_proto_to_file` function too.

Comment (matthewphsmith) [reply flytekit/common/workflow.py:296]: you do you. i'm not sure if fully deprecating the register call is worth the pain though. Adding serialize is a nice option, but I don't think I see the value in removing the existing registration guts and then dealing with the migration.

File: flytekit/clis/sdk_in_container/register.py [hunk 4]
```diff
@@ -13,21 +16,31 @@
...
         project, domain, pkgs, version))
 
     # m = module (i.e. python file)
     # k = value of dir(m), type str
     # o = object (e.g. SdkWorkflow)
+    loaded_entities = []
     for m, k, o in iterate_registerable_entities_in_order(pkgs):
         name = _utils.fqdn(m.__name__, k, entity_type=o.resource_type)
-
+        _logging.debug("Found module {}\n   K: {} Instantiated in {}".format(m, k, o._instantiated_in))
```
Comment (katrogan) [inline flytekit/clis/sdk_in_container/register.py:28]: Did you mean to leave this debug and click echo calls below?

File: flytekit/clis/sdk_in_container/serialize.py [hunk 5]
```diff
@@ -1,50 +1,132 @@
...
+    """
+    :param Text project:
+    :param Text domain:
+    :param list[Text] pkgs:
+    :param Text version:
+    :param Text folder:
+
+    :return:
+    """
+    # m = module (i.e. python file)
```
Comment (katrogan) [inline flytekit/clis/sdk_in_container/serialize.py:32]: could just use descriptive variable names in the first place to avoid these comments :P

File: flytekit/clis/sdk_in_container/serialize.py [hunk 6]
```diff
@@ -1,50 +1,132 @@
...
+        serialized = entity.serialize()
+        fname_index = str(i).zfill(zero_padded_length)
+        fname = '{}_{}.pb'.format(fname_index, entity._id.name)
+        click.echo('  Writing {} to\n    {}'.format(entity._id, fname))
+        _write_proto_to_file(serialized, fname)
+
+        # Not everything serialized will necessarily have an identifier field in it, even though some do (like the
+        # TaskTemplate). To be more rigorous, we write an explicit identifier file that reflects the choices (like
+        # project/domain, etc.) made for this serialize call. We should not allow users to specify a different project
+        # for instance come registration time, to avoid mismatches between potential internal ids like the TaskTemplate
```
Comment (katrogan) [inline flytekit/clis/sdk_in_container/serialize.py:113]: i think admin populates the id in the task template anyways

File: flytekit/clis/sdk_in_container/register.py (hunk 4 above)
Comment (wild-endeavor) [reply flytekit/clis/sdk_in_container/register.py:28]: sorry didn't see this, yeah i did, it's helpful maybe one day.

File: flytekit/clis/sdk_in_container/serialize.py (hunk 6 above)
Comment (wild-endeavor) [reply flytekit/clis/sdk_in_container/serialize.py:113]: true.
